src/tablet/tablet_impl.cc

```diff
@@ -2968,6 +3038,131 @@ int TabletImpl::LoadTableInternal(uint32_t tid, uint32_t pid, std::shared_ptr<::
+        }
+        std::shared_ptr<Snapshot> snapshot = GetSnapshot(tid, pid);
+        if (!snapshot) {
+            PDLOG(WARNING, "snapshot with tid %u and pid %u does not exist", tid, pid);
+            break;
+        }
+        std::shared_ptr<LogReplicator> replicator = GetReplicator(tid, pid);
+        if (!replicator) {
+            PDLOG(WARNING, "replicator with tid %u and pid %u does not exist", tid, pid);
+            break;
+        }
+        {
+            std::lock_guard<SpinMutex> spin_lock(spin_mutex_);
+            table->SetTableStat(::openmldb::storage::kLoading);
+        }
+        uint64_t latest_offset = 0;
+        std::string binlog_path = table_path + "/binlog/";
+        ::openmldb::storage::Binlog binlog(replicator->GetLogPart(), binlog_path);
+        if (binlog.RecoverFromBinlog(table, snapshot_offset, latest_offset)) {
+            table->SetTableStat(::openmldb::storage::kNormal);
+            replicator->SetOffset(latest_offset);
+            replicator->SetSnapshotLogPartIndex(snapshot->GetOffset());
+            replicator->StartSyncing();
+            disk_table->SetOffset(latest_offset);
+            table->SchedGc();
+            gc_pool_.DelayTask(FLAGS_disk_gc_interval * 60 * 1000,
+                               boost::bind(&TabletImpl::GcTable, this, tid, pid, false));
+            io_pool_.DelayTask(FLAGS_binlog_sync_to_disk_interval,
+                               boost::bind(&TabletImpl::SchedSyncDisk, this, tid, pid));
+            task_pool_.DelayTask(FLAGS_binlog_delete_interval,
+                                 boost::bind(&TabletImpl::SchedDelBinlog, this, tid, pid));
+            PDLOG(INFO, "load table success. tid %u pid %u", tid, pid);
+            MakeSnapshotInternal(tid, pid, 0, std::shared_ptr<::openmldb::api::TaskInfo>());
+            std::string old_data_path = table_path + "/old_data";
+            if (::openmldb::base::IsExists(old_data_path)) {
+                if (!::openmldb::base::RemoveDir(old_data_path)) {
+                    PDLOG(WARNING, "remove dir failed. tid %u pid %u path %s", tid, pid, old_data_path.c_str());
+                    break;
+                }
+            }
+            if (task_ptr) {
+                std::lock_guard<std::mutex> lock(mu_);
+                task_ptr->set_status(::openmldb::api::TaskStatus::kDone);
+                return 0;
+            }
+        } else {
+            DeleteTableInternal(tid, pid, std::shared_ptr<::openmldb::api::TaskInfo>());
+        }
+    } while (0);
+    std::string db_root_path;
+    bool ok = ChooseDBRootPath(tid, pid, table_meta.storage_mode(), db_root_path);
+    if (ok) {
+        std::string table_path = db_root_path + "/" + std::to_string(tid) + "_" + std::to_string(pid);
+        std::string old_data_path = table_path + "/old_data";
+        std::string data_path = table_path + "/data";
+        if (::openmldb::base::IsExists(data_path)) {
+            if (!::openmldb::base::RemoveDir(data_path)) {
```

Comment:
you do not know what's the error here? What if error happens before `rename` snapshot to `data_path`?

here also do not `remove`. should `rename` back to snapshot

Reply:
It's a bit complicated here. Maybe if error happens, we just leave as it is. Do nothing to revert.